test_requests.py

```diff
@@ -411,6 +416,21 @@ def test_DIGESTAUTH_QUOTES_QOP_VALUE(self):
         r = requests.get(url, auth=auth)
         assert '"auth"' in r.request.headers['Authorization']
 
+    def test_DIGESTAUTH_THREADED(self):
```

Comment:
I've given some thought to (and sleep over) this - this is as far as I reached:
* Using `threading.Event` to force the coordination of threads to deterministically break them may be feasible.
* It does however imply that such code (test/instrumentation) needs to be within/inline with the code to be tested: this seems to go against the basic principle of testing - separating test code from the code being tested.
* A much more complex alternative (but separate from the code being tested) is to use/build some kind of "clever-testing-thread-runner" that forces its thread pool to behave in a controlled, deterministic fashion regardless of OS scheduling: this might need to hook into lower level runtime/debug hooks or some such... Not being aware of low-level Python runtime internals, I can't even tell if this is currently possible or if it might need extra support there.
* Last, but not least, regarding the test @tardyp shared: given that it is not guaranteed to fail, I wonder if it should be kept -- will it provide a false sense of security? or is it better than having nothing in place, given that the odds of it resulting in a "false success" are very low?
* Additionally, regarding hitting the network: by skimming through `test_requests.py` it seems lots of other tests do it - shouldn't this (admitedly important) challenge be addressed at a more general level? (it seems this can be minimized by running a local instance of httpbin and setting the `HTTPBIN_URL` environment var, which is, again, better than no choice at all)

I'd love to be wrong about the multi-thread testing: if so, please point out my mistakes.
Thanks